### ISSUE TITLE
presigning would be smarter to support

### ISSUE THREAD

**Issue by sebringj** [benjreinhart/react-native-aws3#32]
Currently, I don't need to put my credentials onto any phone because I use presigning from a network request that gives temporary permission to the person uploading with a namespaced bucket. That is a smarter approach in the long run.

Thanks for the code, will just modify it to use signing then do a PR if you want it, take it.

**Comment by BaptWaels** [comment]
It would be great ! 👍  Thanks

**Comment by benjreinhart** [comment]
Yes, it is a better approach, at least in terms of security. I have supplied my thoughts on this in prior closed issues, mainly that I use IAM roles to create a user with as few privileges as possible so that any credential leak would have as limited impact as possible. 

Specifically, I have a role that has only PutObject/PutObjectAcl to one sub directory in my S3 bucket. As far as I can tell, this means absolute worse case scenario is that the credentials could be used to overwrite existing files in that particular folder *if* they knew the file names (file names in my bucket have timestamps, UUIDs or both). Though it's possible I'm missing something else.

So given that, I'm not particularly concerned at the moment. That being said, I do agree that in general, this is still probably not a very safe stance to take on the matter. I built it this way so that I didn't have to deal with the complexity of making requests to a server first to generate the policies and handling any errors that could arise from those network requests. In other words, I chose simplicity/perf over potential security issues when I original built this.

All that being said, I agree this should be supported. I'm pretty sure one of the main reasons for this library having as many forks as it does is because of people who wanted this functionality. I haven't done this yet because I haven't been in React Native world lately it'll take me some time to get this setup and tested to make sure it'll work correctly. I also wanted to have tests for this library before I kept making more changes, which I finally got around to doing this past weekend. So hopefully I'll have some time to get this rolled out shortly.

**Comment by sebringj** [comment]
I ended up making something that does this and it is pretty simple. https://github.com/sebringj/siggy but your code was helpful seeing the specific AJAX post setting for "upload" that I was missing to get progress.

**Comment by benjreinhart** [comment]
Cool, thanks.

This library currently only uses the browser [based upload approach](http://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-UsingHTTPPOST.html) but I would like to switch over to the standard [AWS Signature v4](http://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-authenticating-requests.html) signing which is more flexible. I have a branch on this repo from a while back that was attempting to get that working but ran out of time before I got it fully working. That would allow us to get/put/delete objects rather than only upload. When I have that working, I can also extract the policy generation module which is already pretty standalone within this repo, and then it could be used on the server with node.

**Comment by sebringj** [comment]
thanks, that sounds good.

**Comment by broskoski** [comment]
+1

**Comment by nedmac101** [comment]
+1
